File: Algorithm_39_Combination Sum.java
// 思路：先对candidates排序，保证其从小到大排列，然后使用DFS
// 用curIdx记录当前DFS的起始下标，cur记录当前已选择的所有数之和
// 从curIdx开始，若candidates中当前元素num + cur == target，则已经找到了一组满足要求的combination
// 将这组combination存放到res中并return，无需继续循环下去，因为之后的元素一定比当前num更大，不可能满足和为target
// 若当前num + cur < target，则把num加入curNums中，cur = cur + num，继续DFS

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    private List<List<Integer>> result = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (candidates == null || candidates.length == 0) {
            return result;
        }
        Arrays.sort(candidates);
        DFS(candidates, target, 0, new ArrayList<>());
        return result;
    }
    private void DFS(int[] candidates, int target, int index, List<Integer> cur) {
        if (target == 0) {
            result.add(new ArrayList<>(cur));
            return;
        }
        if (candidates[index] > target) {
            return;
        }
        for (int i = index; i < candidates.length; i++) {
            cur.add(candidates[i]);
            DFS(candidates, target - candidates[i], i, cur);
            cur.remove(cur.size() - 1);
        }
    }
}

File: Algorithm_39_Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        self.result = []
        self.DFS(candidates, len(candidates) - 1, target, [])
        return self.result

    def DFS(self, candidates: List[int], pos: int, target, chosen: List[int]) -> None:
        if target == 0:
            self.result.append(chosen[:])
            return
        if target >= candidates[pos]: # can use candidates[pos]
            chosen.append(candidates[pos])
            self.DFS(candidates, pos, target - candidates[pos], chosen)
            chosen.pop() # remove the appended num
        if pos > 0:
            self.DFS(candidates, pos - 1, target, chosen) # not use candidates[pos]